### PR TITLE
fix(parse): avoid panic when cfg wrapper by bracket under `capture-cfg` mode

### LINKED ISSUE #87577: ICE: Should not have unglued last token with cfg attr
### Code

```Rust
#[derive(Debug)]      
struct S<#[cfg(feature = "alloc")] N: A<T>> {}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.56.0-nightly (2faabf579 2021-07-27)
binary: rustc
commit-hash: 2faabf579323f5252329264cc53ba9ff803429a3
commit-date: 2021-07-27
host: x86_64-unknown-linux-gnu
release: 1.56.0-nightly
LLVM version: 12.0.1
```

First fails in `rustc 1.53.0-nightly (a86612456 2021-04-11)`.

Compiles successfully in `rustc 1.53.0-nightly (a836d9b64 2021-04-10)`.

Perhaps related to #82608 which was merged on that date and appears when searching "unglued."

### Error output

```
thread 'rustc' panicked at 'Should not have unglued last token with cfg attr', compiler/rustc_parse/src/parser/attr_wrapper.rs:371:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.56.0-nightly (2faabf579 2021-07-27) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental -C link-arg=-fuse-ld=lld --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `unglued`
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'Should not have unglued last token with cfg attr', compiler/rustc_parse/src/parser/attr_wrapper.rs:371:13
stack backtrace:
   0: std::panicking::begin_panic
   1: rustc_parse::parser::generics::<impl rustc_parse::parser::Parser>::parse_generic_params
   2: rustc_parse::parser::generics::<impl rustc_parse::parser::Parser>::parse_generics
   3: rustc_parse::parser::item::<impl rustc_parse::parser::Parser>::parse_item_kind
   4: rustc_parse::parser::item::<impl rustc_parse::parser::Parser>::parse_item_common
   5: rustc_parse::parser::item::<impl rustc_parse::parser::Parser>::parse_item
   6: rustc_builtin_macros::cfg_eval::cfg_eval
   7: <rustc_builtin_macros::derive::Expander as rustc_expand::base::MultiItemModifier>::expand
   8: rustc_expand::expand::MacroExpander::fully_expand_fragment
   9: rustc_expand::expand::MacroExpander::expand_crate
  10: rustc_session::utils::<impl rustc_session::session::Session>::time
  11: rustc_interface::passes::configure_and_expand
  12: rustc_interface::queries::Queries::expansion
  13: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  14: rustc_span::with_source_map
  15: rustc_interface::interface::create_compiler_and_run
  16: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

</p>
</details>

### PR DESCRIPTION
**Pull request by bvanjoi**
Fixes #87577

The parser expected to encounter the `Token::Gt` but instead encountered `BinOp(shr)` when parsing `A<T>` during macro expand. As a result, the `self.token_cursor.break_last_token = true` in `break_and_eat` had been executed. This caused the parser to encounter `assert!(!self.token_cursor.break_last_token, xxx)` in `﻿collect_tokens_trailing_token` under `cpature-cfg` mode.

To prevent this issue, an extra restriction was added to the `assert` statement. By the way, I think the better solution may be to simply remove the `assert` altogether.

### CODE REVIEW

**Comment by bvanjoi** [comment]
Hi @Aaron1011, I see that you assigned this task to yourself, but since it has been two years since the original and I try to solve it. 

By the way, I will appreciate it if you can help to review this change from the author view?

File: compiler/rustc_parse/src/parser/attr_wrapper.rs
Comment (davidtwco) [inline compiler/rustc_parse/src/parser/attr_wrapper.rs:370]: I'm not too familiar with this code but you could look into why the token isn't broken? It strikes me as unexpected that we'd give the derive a `BinOp(shr)` here instead of two `Gt` tokens?

Comment (bvanjoi) [reply compiler/rustc_parse/src/parser/attr_wrapper.rs:370]: > why the token isn't broken?

This is intentional:

https://github.com/rust-lang/rust/blob/8882507bc7dbad0cc0548204eb8777e51ac92332/compiler/rustc_parse/src/lexer/tokentrees.rs#L59-L61

**Comment by davidtwco** [comment]
I'm not familiar enough with this part of the compiler, I'll let Aaron take a look.

r? @Aaron1011

**Comment by Aaron1011** [comment]
Sorry for the delay in getting to this.

I don't think this is right - that assertion is to ensure that we correctly capture tokens, since capturing an unglued token didn't work at the time I wrote this.

Can you add a test using an attribute that prints/modifies the tokens, and make sure that it sees the correct tokens with this change?

**Comment by bvanjoi** [comment]
> I don't think this is right 

However, it appears that there is no appropriate position for updating ﻿`self.break_last_token` during macro expansion.

> Can you add a test using an attribute that prints/modifies the tokens

Since attributes cannot print the result of the macro expansion, I've used hir printing to ensure correctness.

**Comment by wesleywiser** [comment]
Hi @Aaron1011, it looks like this is ready for another round of review when you get a chance. Cheers!

File: compiler/rustc_parse/src/parser/attr_wrapper.rs
Comment (estebank) [inline compiler/rustc_parse/src/parser/attr_wrapper.rs:367]: @petrochenkov this change is so that we no longer ICE on the face of

```rust
struct S<#[cfg(feature = "alloc")] N: A<T>>;
                                         ^^ these currently get split and cause the assertion to fail unconditionally
```

I believe that this change is reasonable. Is there *any* reason we shouldn't land this PR as is? Put another way, are there any non-local knock down effects from this change that come to mind?

**Review by estebank — APPROVED** [review]

File: compiler/rustc_parse/src/parser/attr_wrapper.rs
Comment (apiraino) [reply compiler/rustc_parse/src/parser/attr_wrapper.rs:367]: gentle ping @petrochenkov for this comment. thanks.

**Comment by petrochenkov** [comment]
This needs to be tested with proc macros working on tokens, not with `Debug` which works on AST.
`print_derive` from `tests\ui\proc-macro\auxiliary\test-macros.rs` can be used for testing.

The derive should see the structure with the `#[cfg(feature = "alloc")]` part configured out, that means `struct S<>;`.
@rustbot author

**Comment by JohnCSimon** [comment]
@bvanjoi 
ping from triage - can you post your status on this PR? This PR has not received an update in a few months.

**Comment by JohnCSimon** [comment]
@bvanjoi 
Ping from triage: I'm closing this due to inactivity as it hasn't been touched in months
Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Comment by JohnCSimon** [comment]
@bvanjoi triage: can you post your status on this?

**Comment by bvanjoi** [comment]
I still haven't found the right solution, so I'm closing this.

**Pull request closed without merging**
